gtsfm/frontend/correspondence_generator/mast3r_correspondence_generator.py

```diff
@@ -0,0 +1,341 @@
+import mast3r.cloud_opt.sparse_ga as mast3r_ga
+
+import numpy as np
+import torch
+import torchvision.transforms as tvf
+from dask.distributed import Client, Future
+
+from gtsfm.common.image import Image
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.utils import images as image_utils
+from gtsfm.utils import logger as logger_utils
+from gtsfm.frontend.correspondence_generator.correspondence_generator_base import (
+    CorrespondenceGeneratorBase,
+)
+
+
+logger = logger_utils.get_logger()
+_MODEL_PATH = str(
+    Path(__file__).resolve().parent.parent.parent.parent
+    / "thirdparty"
+    / "mast3r"
+    / "checkpoints"
+    / "MASt3R_ViTLarge_BaseDecoder_512_catmlpdpt_metric.pth"
+)
+GRID_DIVISOR = 16
+PATCH_MULTIPLIER = 8
+
+def resize_image(img: Image, long_edge_size: int) -> Image:
+    """Resizes image such that longest edge is equal to long_edge_size.
+
+    Args:
+        img: The input image to be resized.
+        long_edge_size: The desired size for the longest edge of the image.
+
+    Returns:
+        The resized image.
+    """
+    max_size = max(img.height, img.width)
+    ratio = float(long_edge_size) / max_size
+    new_height = int(img.height * ratio)
+    new_width = int(img.width * ratio)
+    new_image = image_utils.resize_image(img, new_height, new_width)
+    return new_image
+
+
+def preprocess_image(image: Image, img_id: int, device) -> Tuple[Dict[str, Any], Tuple[int, int], float]:
+    """Preprocesses image for MASt3R model input, including resizing and normalization.
+
+    Args:
+        image: The input image to be preprocessed.
+        img_id: An identifier for the image.
+        device: The device (e.g., 'cpu' or 'cuda') to which the image tensor should be moved.
+
+    Returns:
+        A tuple containing:
+        - A dictionary with the preprocessed image tensor, its true shape, id, and instance string.
+        - A tuple (x, y) representing the start coordinates of the crop applied to the image.
+        - The scaling factor applied to the image during resizing.
+    """
+    H1, _, _ = image.shape
+
+    # resize to longest edge being `size`.
+    img = resize_image(image, long_edge_size=512).value_array
+    H, W, _ = img.shape
+    scale_factor = float(H) / H1
+
+    cx, cy = W // 2, H // 2
+    halfw, halfh = ((2 * cx) // GRID_DIVISOR) * PATCH_MULTIPLIER, ((2 * cy) // GRID_DIVISOR) * PATCH_MULTIPLIER
+    crop_start = (cx - halfw, cy - halfh)
+    img = img[crop_start[1] : cy + halfh, crop_start[0] : cx + halfw]
+    ImgNorm = tvf.Compose([tvf.ToTensor(), tvf.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
+
+    return (
+        dict(
+            img=ImgNorm(img)[None].to(device),
+            true_shape=np.int32([img.shape[:2]]),
+            idx=img_id,
+            instance=str(img_id),
+        ),
+        crop_start,
+        scale_factor,
+    )
+
+
+def symmetric_inference(model, img1, img2, device):
+    """Performs symmetric inference using the MASt3R model on a pair of images.
```

Comment:
Could you add a slightly more detailed description of what symmetric inference is?

Reply:
Still think this would be useful